### DISCUSSION TITLE
Inserting PDN takes a lot of time

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#2790] [General]
Is it expected that inserting PDN (last part of make floorplan) should take a lot of time for designs that start to approach 1mm^2 in ASAP7?

It would be helpful if "make floorplan" was faster, because then it is quicker to iterate on the floorplan, after which the build can normally be left alone without human interaction for longer runs.

I ran it in the debugger and it looks like a lot of time is spent building this ShapeTree: https://github.com/The-OpenROAD-Project/OpenROAD/blob/b99d7d2c3db8b72a3fc35da9bd6c3a203c3b8306/src/pdn/src/shape.cpp#L637 

Afterwards, this line seems to almost always return false https://github.com/The-OpenROAD-Project/OpenROAD/blob/b99d7d2c3db8b72a3fc35da9bd6c3a203c3b8306/src/pdn/src/shape.cpp#L640) 

I made a few fleeting vain attempts to rewrite the code, such that the common case, https://github.com/The-OpenROAD-Project/OpenROAD/blob/b99d7d2c3db8b72a3fc35da9bd6c3a203c3b8306/src/pdn/src/shape.cpp#L627, returning false, would not require building the ShapeTree.

**Comment by maliberty** [comment]
There could be a lot of factors.  Are you using the asap7 default pdn?
@gadfort

**Reply by oharboe** [reply]
Yes.

**Comment by oharboe** [comment]
Still running, 1 hr 35 minutes...

```
01:35:49 /OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad -exit -no_init ./scripts/pdn.tcl -metrics ./logs/asap7/doohickey/base/2_6_pdn.json
```

**Comment by maliberty** [comment]
In asap7 the m1/m2 stripes introduce a ton of vias.  @gadfort could help skip the add_pdn_connect for M1/M2 or would that produce an error.  It wouldn't matter much for prototyping whether we have the vias or not.

**Comment by gadfort** [comment]
@oharboe @maliberty it shouldn't take that long, there has always been some inefficiency when building those obstruction sets and maybe it's a good idea to revisit how that is done and find a better solution that is faster.

**Reply by maliberty** [reply]
@gadfort for the m1/m2 via stapling you could easily treat that as one big obstruction rather than all the individual vias.

**Comment by oharboe** [comment]
@maliberty @gadfort Example of ```[INFO PDN-0001] Inserting grid: top``` taking at least 12 hours, after which I stopped it. https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/779

**Reply by oharboe** [reply]
@maliberty @gadfort A stab in the dark at fixing it. Runs much faster, but does it still do the right thing? https://github.com/The-OpenROAD-Project/OpenROAD/pull/2795

**Reply by gadfort** [reply]
Thanks for isolating the slow step could you please upload a testcase so I can try out my possible fix?

**Reply by oharboe** [reply]
Here you go: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/779/commits

**Reply by gadfort** [reply]
So that doesn't tell me which design also when I attempt them I get placement errors.

**Reply by gadfort** [reply]
@oharboe could you please provide a runnable testcase. I don't want to try to debug multiple issues in that pull request.

**Reply by oharboe** [reply]
Use this pull request, https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/779/, then run:

```
make DESIGN_CONFIG=designs/asap7/mock-very-large-array/config.mk
```

This will first build the "Element" macro, then the top level design will be built. It is when the PDN is inserted in the top level build you will see the speedup of my hack.

**Reply by gadfort** [reply]
Element is failing placement for me. So I can't replicate, please generate a make issue_pdngen for the slow step so I don't need to figure out why placement is not working right.

**Reply by oharboe** [reply]
I'll knock this test-case into shape so that it passes on Jenkins: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/782

**Reply by gadfort** [reply]
@oharboe I think I managed to trick things to run, atleast enough to try it out. My preference (for future reference) would be to use the issue feature instead of a PR on ORFS, it just makes things harder not easier.

**Reply by oharboe** [reply]
Thank you. Noted.

PDN insertion is one important thing to fix that I wanted to cover by the test case above. I'm hoping to get the test case into master, if it isn't regression tested, it will bit-rot sort of thinking.

**Comment by oharboe** [comment]
PDN insertion is MUCH faster than it used to be, but it can still be an issue.
